### PR TITLE
Make exchange client use its own memory pool plus bug fixes

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Currently, the exchange client of the exchange node from a pipeline
uses the first instantiated operator's memory pool which is a little bit
hack and is not good in memory usage counting. This PR changes to
create a dedicated memory pool for exchange client's memory usage
which provides better memory usage counting as well as ease memory
leaking related bug analysis. We also create a dedicated memory pool
for each merge exchange source.

Fixes a bug in query plan node memory pool creation. The current code
creates a query plan node memory pool for each driver which is caused
by the bug in the code. Add a unit test to verify the memory pool structure

Also fixes operator memory pool node's name to make it unique by adding
driver id in naming.

### CODE REVIEW

**Comment by tanjialiang** [comment]
Thanks @xiaoxmeng for the refactoring. Things look cleaner now. One concern though. Currently in Presto look-up page we show memory metrics on operator level. The newly changed code put Exchange memory usage in a direct plan node child pool, decoupling the relation from an operator. Hence while debugging we will see ExchangeOperator memory usage being 0 always. I wonder if we should also somehow have that same pool populated in operator as well? Or do we have some other solutions for this?

**Comment by tanjialiang** [comment]
> Thanks @xiaoxmeng for the refactoring. Things look cleaner now. One concern though. Currently in Presto look-up page we show memory metrics on operator level. The newly changed code put Exchange memory usage in a direct plan node child pool, decoupling the relation from an operator. Hence while debugging we will see ExchangeOperator memory usage being 0 always. I wonder if we should also somehow have that same pool populated in operator as well? Or do we have some other solutions for this?

Actually I think it should be populated properly as the population is on plan node level. Thanks for the refactoring!

**Review by tanjialiang — APPROVED** [review]

**Comment by xiaoxmeng** [comment]
> > Thanks @xiaoxmeng for the refactoring. Things look cleaner now. One concern though. Currently in Presto look-up page we show memory metrics on operator level. The newly changed code put Exchange memory usage in a direct plan node child pool, decoupling the relation from an operator. Hence while debugging we will see ExchangeOperator memory usage being 0 always. I wonder if we should also somehow have that same pool populated in operator as well? Or do we have some other solutions for this?
> 
> Actually I think it should be populated properly as the population is on plan node level. Maybe run a query to verify? Thanks for the refactoring!

Sure, will do. Thanks for pointing out!

File: velox/exec/Exchange.h [hunk 1]
```diff
@@ -298,8 +298,12 @@ class ExchangeClient {
  public:
   static constexpr int32_t kDefaultMinSize = 32 << 20; // 32 MB.
 
-  explicit ExchangeClient(int destination, int64_t minSize = kDefaultMinSize)
+  ExchangeClient(
+      int destination,
+      memory::MemoryPool* FOLLY_NONNULL pool,
```
Comment (mbasmanova) [inline velox/exec/Exchange.h:303]: remove FOLLY_NONNULL

File: velox/exec/Exchange.h [hunk 2]
```diff
@@ -338,10 +340,10 @@ class ExchangeClient {
 
  private:
   const int destination_;
+  memory::MemoryPool* const FOLLY_NONNULL pool_;
```
Comment (mbasmanova) [inline velox/exec/Exchange.h:343]: remove FOLLY_NONNULL

File: velox/exec/MergeSource.cpp [hunk 3]
```diff
@@ -119,10 +119,10 @@ class MergeExchangeSource : public MergeSource {
   MergeExchangeSource(
       MergeExchange* mergeExchange,
       const std::string& taskId,
-      int destination)
+      int destination,
+      memory::MemoryPool* FOLLY_NONNULL pool)
```
Comment (mbasmanova) [inline velox/exec/MergeSource.cpp:123]: remove FOLLY_NONNULL

File: velox/exec/MergeSource.cpp [hunk 4]
```diff
@@ -181,8 +181,8 @@ class MergeExchangeSource : public MergeSource {
   }
 
  private:
-  MergeExchange* mergeExchange_;
-  std::unique_ptr<ExchangeClient> client_;
+  MergeExchange* const mergeExchange_;
+  std::shared_ptr<ExchangeClient> client_;
```
Comment (mbasmanova) [inline velox/exec/MergeSource.cpp:185]: Why is this a shared_ptr?

File: velox/exec/MergeSource.h [hunk 5]
```diff
@@ -39,7 +39,8 @@ class MergeSource {
   static std::shared_ptr<MergeSource> createMergeExchangeSource(
       MergeExchange* mergeExchange,
       const std::string& taskId,
-      int destination);
+      int destination,
+      memory::MemoryPool* FOLLY_NONNULL pool);
```
Comment (mbasmanova) [inline velox/exec/MergeSource.h:43]: remove FOLLY_NONNULL

File: velox/exec/Task.cpp [hunk 6]
```diff
@@ -195,22 +195,45 @@ Task::getOrAddNodePool(const core::PlanNodeId& planNodeId) {
...
+  childPools_.push_back(nodePool->addChild(fmt::format(
+      "op.{}.{}.{}.{}", planNodeId, pipelineId, driverId, operatorType)));
+  return childPools_.back().get();
+}
+
+velox::memory::MemoryPool* Task::addMergeSourcePool(
+    const core::PlanNodeId& planNodeId,
+    uint32_t pipelineId,
+    uint32_t sourceId) {
+  std::lock_guard<std::mutex> l(mutex_);
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:224]: Why don't we use the lock in Task::addOperatorPool ?

File: velox/exec/Task.cpp [hunk 7]
```diff
@@ -195,22 +195,45 @@ Task::getOrAddNodePool(const core::PlanNodeId& planNodeId) {
...
+  auto* nodePool = getOrAddNodePool(planNodeId);
+  childPools_.push_back(nodePool->addChild(fmt::format(
+      "mergeExchangeClient.{}.{}.{}", planNodeId, pipelineId, sourceId)));
+  return childPools_.back().get();
+}
+
+velox::memory::MemoryPool* Task::addExchangeClientPool(
+    const core::PlanNodeId& planNodeId,
+    uint32_t pipelineId) {
+  auto* nodePool = getOrAddNodePool(planNodeId);
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:234]: Why don't we use a lock here?

File: velox/exec/Task.h [hunk 8]
```diff
@@ -264,20 +264,23 @@ class Task : public std::enable_shared_from_this<Task> {
...
   velox::memory::MemoryPool* FOLLY_NONNULL addOperatorPool(
       const core::PlanNodeId& planNodeId,
       int pipelineId,
+      uint32_t driverId,
       const std::string& operatorType);
 
+  /// Creates new instance of MemoryPool for a merge source in a
+  /// MergeExchangeNode, stores it in the task to ensure lifetime and returns a
+  /// raw pointer.
+  velox::memory::MemoryPool* FOLLY_NONNULL addMergeSourcePool(
```
Comment (mbasmanova) [inline velox/exec/Task.h:279]: remove FOLLY_NONNULL

File: velox/exec/Task.h [hunk 9]
```diff
@@ -509,6 +512,18 @@ class Task : public std::enable_shared_from_this<Task> {
   static void testingWaitForAllTasksToBeDeleted(uint64_t maxWaitUs = 3'000'000);
 
  private:
+  // Creates new instance of MemoryPool for a plan node, stores it in the task
+  // to ensure lifetime and returns a raw pointer.
+  memory::MemoryPool* FOLLY_NONNULL
```
Comment (mbasmanova) [inline velox/exec/Task.h:517]: remove FOLLY_NONNULL

File: velox/exec/Task.h [hunk 10]
```diff
@@ -509,6 +512,18 @@ class Task : public std::enable_shared_from_this<Task> {
...
+  getOrAddNodePool(const core::PlanNodeId& planNodeId);
+
+  // Creates new instance of MemoryPool for the exchange client of an
+  // ExchangeNode in a pipeline, stores it in the task to ensure lifetime and
+  // returns a raw pointer.
+  velox::memory::MemoryPool* FOLLY_NONNULL addExchangeClientPool(
```
Comment (mbasmanova) [inline velox/exec/Task.h:523]: remove FOLLY_NONNULL

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Thanks.

File: velox/exec/Task.cpp (hunk 6 above)
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:224]: addOperatorPool is called on operator ctor which doesn't or can't take a look as it is called from createDriverLocked

File: velox/exec/Task.cpp (hunk 7 above)
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:234]: This is called from Task::start which will have lock protection in the caller (race condition fix diff). I will comment this in a followup cleanup when both diffs landed.

File: velox/exec/MergeSource.cpp (hunk 4 above)
Comment (xiaoxmeng) [reply velox/exec/MergeSource.cpp:185]: No need. Changed it to unique_ptr. Thanks!

**Pull request closed without merging**